### ISSUE TITLE
Research 'flaky test handling' process

### ISSUE THREAD

**Issue by rokonec** [dotnet/arcade#6132]
In order to properly implement Phase I of #5963, capturing data, I need to understand the problem and design a high level process handling it.

- [x] study [Resiliency in our infrastructure.docx](https://microsoft-my.sharepoint.com/:w:/r/personal/jaredpar_microsoft_com/_layouts/15/guestaccess.aspx?e=0nRQWy&share=EfdbxSdPRKdKnYTd84gxdeUBYuB5AYiKVcbTQcXP2SdzdQ) and related sources
  - https://martinfowler.com/articles/nonDeterminism.html
  - https://testing.googleblog.com/2016/05/flaky-tests-at-google-and-how-we.html
  - https://medium.com/scopedev/how-can-we-peacefully-co-exist-with-flaky-tests-3c8f94fba166
  - [1ES Flaky Test Management](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdev.azure.com%2Fmseng%2F1ES%2F_wiki%2Fwikis%2F1ES%2520Test%2520Docs%2F12782%2FFlaky-Test-Management-System-Common&data=02%7C01%7Cjaredpar%40microsoft.com%7C0e49808ec29943e1b34c08d83fbecd06%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637329435997868814&sdata=mplu2mJsG1riX9IAzJ7wddAqLJVeYHweyTM9%2BMA8CEY%3D&reserved=0)
  - https://docs.microsoft.com/en-us/azure/devops/pipelines/test/flaky-test-management?view=azure-devops
- [x] study Kusto query language and optimizations
- [x] analyze our Kusto stored telemetry 'TestResults' and try to use it to identify flaky tests
- [ ] discuss the problem with our partners (Jared, Doug, ...)
- [ ] validate the proposed high level flaky-test-handling process (Jared, Doug, Chris, Tomas, ...)

**Comment by rokonec** [comment]
## Current Kusto telemetry

![image](https://user-images.githubusercontent.com/25249058/92604569-f19d3880-f2b0-11ea-8ab0-98b369836f21.png)

Data about **every** test run are currently ingested into **TestResults** table and **kept for 60 days**. 
This is enormous amount of data ~ 60E9 rows, and even Kusto have reached its limits when processing some even relatively simple queries like joining it with Jobs table. 
I am not sure if all tests we run are ingested into this table, but since **Runtime** test consumes about **95%** of this data, the repository which suffer most from flaky tests seems to be covered.

## Useful queries (mostly working)

### Flaky suspicious jobs
In theory merge to master shall be done when all tests on PR passed. So every main/master builds failed for test purposes is suspicious of flake-fail. In reality master build often tests more than PR to improve PR workflow experience and save resources.

One way or another, dealing with 100+ failed **main** builds per day is a considerable effort.

```KQL
Jobs
| where Started > ago(30d) 
| where Branch has 'master' or Branch has 'main'
| where TestsFail > 0
```

![image](https://user-images.githubusercontent.com/25249058/92606779-9e78b500-f2b3-11ea-9420-36903e3affc1.png)

### Particular test flakiness analysis
In an attempt to guess if given test is flaky I have came up with following flakiness ratio *r*:
changes = number of changes from passing to failing, lets presume that two consequent run of same test runs fail for same reason
failures = number of failed test runs
t = measurement time

`r = (changes/t) / (failures/t) => changes/failures`

r = 1 means highest probability of test be be flaky as every its every failure was followed by pass. It presumes that bug which makes it into master will be fixed after some significant time during which other master builds will fail.

`fr` is simple fail rate = failures/total

```KQL
let t = 'JIT_Regression._CLR_x86_JIT_V1_M09_5_PDC_b26957_b26957_b26957_';
let m = '_CLR_x86_JIT_V1_M09_5_PDC_b26957_b26957_b26957_sh';
let h = '';
TestResults | where Type == t and Method == m and ArgumentHash == h
| extend at=ingestion_time(), failed=(Result == 'Fail')
| where at > ago(30d)
| order by at asc
| summarize t=make_list(at), failed=make_list(todouble(failed)) by Type, Method, ArgumentHash
| extend changes=series_fir(failed, dynamic([1,-1]))
| mv-expand changes, failed
| summarize changes=countif(changes==1), failed=countif(failed==1), total=count() by Type, Method, ArgumentHash
| extend r=changes/todouble(failed), fr=failed/todouble(total)
| order by r desc
```

![image](https://user-images.githubusercontent.com/25249058/92613598-6d03e780-f2bb-11ea-83d2-dca405ed0c09.png)

This is just prototype and it was not tested yet if this actually correlates to true test flakiness. 

### Tests failure rates
This query can be used to materialize its results for analytics and reporting purposes.

Limit bellow query for particular test or test Type (class) might be also useful.

```KQL
set notruncation;
set maxmemoryconsumptionperiterator=68719476736;
set max_memory_consumption_per_query_per_node=68719476736;
TestResults 
| where ingestion_time() > ago(90d)
| summarize failed=countif(Result == 'Fail'), total=count() by Type, Method, ArgumentHash
| extend fr=failed/todouble(total)
| where fr != 0
| order by fr desc
```

## Recommended Optimization and improvements
To be able to extract data with higher business value I recommend to consider following enhancements to TestResults telemetry:

- Add branch column, this was we can identify rollout tests run or particular PR branch test runs
- Add commit sha-7, so we can identify tests which fails on same commit and thus are likely flaky

To have better performance of common Kusto query patterns I suggest following changes
- denormalize **Jobs** (Project, Repository, Name, Type, TestFail/Pass, JobFail/Pass)
- denormalize **Jobs** Propertis (BuildNumber, BuildDefinition, OS,  Name, Type, TestFail/Pass, JobFail/Pass)
- denormalize **WorkItems** (Status, Queue, MachineName, FriendlyName, tests failure rate)
- Traits column shall contain json dictionary, as Kusto has optimized way to work with json strings as oppose to parse custom string format

## Conclusion
I believe that collecting test run telemetry is great and can be used to obtain some valuable business intelligence and/or improve flaky-test handling process.

**Comment by rokonec** [comment]
## Flaky tests process

I was thinking about two possible processes which could help handling flaky tests friction.

## Simply retry
![image](https://user-images.githubusercontent.com/25249058/92733147-a39f3800-f377-11ea-9e45-017597650184.png)

Pros:
- simple to implement
- simple to understand
- could help a lot during implementing an advanced process

Cons:
- We have to implement back off retry in order to address common lengthy infrastructure flakiness. I think in range of 15 seconds to 5 minutes. Any of those could lead to unacceptable tests prolonging end if not treated, eventual build timeout. We might be forced to implement some global cumulative wait time and stop retrying if exhausted.
- Retry expensive tests might be costly (time and resources)
- Will make new written tests flakiness less obvious. Eventually it could encourage lazy-dev behavior as threshold for problematic test flakiness could drop significantly. Say our current threshold to fix flaky test is 0.0001 fail rate, with just 4 retries we will experience same build failure with tests which inner fail rate 0.1. 10% failure rate should be always addressed but our retry will make such cases less obvious.

**Comment by rokonec** [comment]
I am no longer working on this epic.
